tests/test_boundary_conditions.py

```diff
@@ -221,6 +217,110 @@ def test_fixed_constraint(seed, n_position_constraint, n_director_constraint):
+        translational_constraint_selector=translational_constraint_selector,
+        rotational_constraint_selector=rotational_constraint_selector,
+        _system=test_rod,
+    )
+    pos_indices = rng.choice(
+        test_rod.n_elem, size=num_translational_constraint, replace=False
+    )
+    dir_indices = rng.choice(
+        test_rod.n_elem, size=num_rotational_constraint, replace=False
+    )
+    fixed_rod._constrained_position_idx = pos_indices.copy()
+    fixed_rod._constrained_director_idx = dir_indices.copy()
+
+    test_position_collection = rng.random((3, test_rod.n_elem))
+    test_rod.position_collection = (
+        test_position_collection.copy()
+    )  # We need copy of the list not a reference to this array
+    fixed_rod.constrain_values(test_rod, time=0)
+    test_position_collection[..., pos_indices] = start_position_collection.transpose(
+        (1, 0)
+    )
+
+    assert_allclose(
+        test_position_collection[translational_constraint_selector, :],
+        test_rod.position_collection[translational_constraint_selector, :],
+        atol=Tolerance.atol(),
+    )
+
+    test_velocity_collection = rng.random((3, test_rod.n_elem))
+    test_rod.velocity_collection = (
+        test_velocity_collection.copy()
+    )  # We need copy of the list not a reference to this array
+    test_director_collection = rng.random((3, 3, test_rod.n_elem))
+    test_rod.director_collection = (
+        test_director_collection.copy()
+    )  # We need copy of the list not a reference to this array
+    test_omega_collection = rng.random((3, test_rod.n_elem))
+    test_rod.omega_collection = (
+        test_omega_collection.copy()
+    )  # We need copy of the list not a reference to this array
+    fixed_rod.constrain_rates(test_rod, time=0)
+
+    # test `nb_constrain_translational_rates`
+    if translational_constraint_selector[0]:
+        test_velocity_collection[0, pos_indices] = 0.0
+    if translational_constraint_selector[1]:
+        test_velocity_collection[1, pos_indices] = 0.0
+    if translational_constraint_selector[2]:
+        test_velocity_collection[2, pos_indices] = 0.0
+    assert_allclose(
+        test_velocity_collection, test_rod.velocity_collection, atol=Tolerance.atol()
+    )
+
+    # test `nb_constrain_rotational_rates`
+    # rotate angular velocities to inertial frame
+    omega_collection_lab_frame = _batch_matvec(
+        test_director_collection[..., dir_indices].transpose(1, 0, 2),
+        test_omega_collection[..., dir_indices],
+    )
+    # apply constraint selector to angular velocities in inertial frame
+    omega_collection_not_constrained = (
+        1 - np.expand_dims(rotational_constraint_selector, 1)
+    ) * omega_collection_lab_frame
+    # rotate angular velocities vector back to local frame and apply to omega_collection
+    test_omega_collection[..., dir_indices] = _batch_matvec(
+        test_director_collection[..., dir_indices], omega_collection_not_constrained
+    )
```

Comment:
this is the same code copied from inside the function, hence may not be a valid test. The idea is to have a manufactured solution (based on a fixed set of inputs, which can be a bit trivial) and then test it against the function result. In other words, test the function against the same result but obtained via a different method (could be hand calculated manufactured solution in this case)

Reply:
I slightly adjusted the test code and added a special cases for 
- the directors being equal to the identity matrix in https://github.com/GazzolaLab/PyElastica/pull/143/commits/33d42f1b9e2b7efa3e8e5d05669cc8e6d200bb0c
- the a 90 degree rotation around the z-axis in https://github.com/GazzolaLab/PyElastica/pull/143/commits/4e8c6e212dbf99980d6142ad9728fbb9b12a4ce3